elasticdl/python/worker/task_data_service.py

```diff
@@ -26,44 +32,90 @@ def __init__(
         self._current_task = None
 
     def get_current_task(self):
         return self._current_task
 
+    def report_task_done(self, task_id, err_msg, exec_counters=None):
+        self._worker.report_task_result(
+            task_id, err_msg, exec_counters=exec_counters
+        )
+
+    def handle_next_task(self, err_msg=""):
+        with self._lock:
+            # Reason for keep_poping, batch_size may be large than
+            # task.end - task.start, so we keep poping task until
+            # reported count is less than current data size
+            keep_poping = True
+
+            # Becuase a single batch comes from multiple tasks,
+            # We just report the number of failing records together
+            # with first task
+            if self._failed_record_count != 0:
+                exec_counters = {
+                    TaskExecCounterKey.FAIL_COUNT: self._failed_record_count
+                }
+            else:
+                exec_counters = None
+            while keep_poping:
+                task = self._pending_tasks.popleft()
+                self._reported_record_count -= task.end - task.start
+                self.report_task_done(
+                    task.task_id, err_msg, exec_counters=exec_counters
+                )
+                exec_counters = None  # reset counters
+                if self._pending_tasks:
+                    self._current_task = self._pending_tasks[0]
+                    task_len = (
+                        self._current_task.end - self._current_task.start
+                    )
+                    keep_poping = self._reported_record_count >= task_len
+                else:
+                    break
+            self._failed_record_count = 0
+
     def report_record_done(self, count, err_msg=""):
         """
         Report the number of records in the latest processed batch,
         so TaskDataService knows if some pending tasks are finished
         and report_task_result to the master.
-        self._pending_tasks_with_counts[0][0] is the first pending task,
-        self._pending_tasks_with_counts[0][1] is the number of records
-        in this task.
         """
-        self._reported_record_count += count
-        if (
-            len(self._pending_tasks_with_counts)
-            and self._reported_record_count
-            >= self._pending_tasks_with_counts[0][1]
-        ):
-            with self._lock:
-                while (
-                    len(self._pending_tasks_with_counts)
-                    and self._reported_record_count
-                    >= self._pending_tasks_with_counts[0][1]
-                ):
-                    task, _ = self._pending_tasks_with_counts.pop(0)
-                    self._worker.report_task_result(task.task_id, err_msg)
-                if len(self._pending_tasks_with_counts):
-                    self._current_task = self._pending_tasks_with_counts[0][0]
+        if self._pending_tasks:
```

Comment:
This methdod `report_record_done` can be simplified:
```
def report_record_done(self, count, err_msg=""):
    self._reported_record_count += count
    if err_msg:
            self._failed_record_count += count

    pending_task_finished = False
    if self._pending_tasks:
         task=  self._pending_tasks[0]
         if self._reported_record_count >= task.end - task.start:
             pending_task_finished = True
         if pending_task_finished:
             with self._lock:
                 while  self._pending_tasks and self._reported_record_count >= self._pending_tasks[0].end - self._pending_tasks[0].start:
                  task = self._pending_tasks.pop(0)
                  self._reported_record_count -= task.end - task.start
                  then report_task_done with some err_msg/exec_counters
       
```

Reply:
Updated.